### PR TITLE
Added support for parsing rules and sanitizing HAR files

### LINKED ISSUE #2: Come up with rule file
```
name:
description:
format: <json|xml|txt>
rules:
    <JSONPath|XPath|Regex>:
            description:
            action: <contextual_replace|remove>
```

https://skarlso.github.io/2023/12/01/crd-to-yaml-as-wasm-website/

### PR DESCRIPTION
**Pull request by the-c0d3br34k3r**
## Description
Closes #2 and #3.
Added support for parsing rules and sanitizing HAR files
Rule file template: https://github.com/padaiyal/sanitizer/issues/2#issue-2135617656

## Author's Checklist
- [x] **These changes don't break existing dependants.** If they do, and you suggest to proceed, 
please explain why.

## Reviewer's Checklist
- [ ] **All classes have documentation comments.** - NA
- [ ] **All methods have documentation comments.** - NA
- [ ] **No parameters are hardcoded.** Should be loaded as properties instead. If there needs to be hardcoded properties, explain why. - Some config values are hardcoded. Will be moved to a config file in a #8.
- [ ] **All information to be logged/displayed to the user are internationalized.** If not, explain why. - NA
- [ ] **README.md has been updated if needed.** - NA
- [ ] **pom.xml version is set to the latest date** If not, explain why. - NA

## Assignee's Checklist
- [ ] **All GitHub action checks have passed.** N/A
- [x] **All reviewers have approved.**
- [x] **Relevant documentation has been updated if needed.**
- [ ] **pom.xml version is set to the latest date** If not, explain why. N/A

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1,47 +1,221 @@
...
+    const sanitizeActions = {};
+    for (const path in rules) {
+      const action = rules[path]["action"]
+      if (!supportedActions.includes(action)) {
+        console.warn(`Skipping rule "${rules[path]["description"]}
+                          - Unsupported action "${action}.`);
+        continue;
+      }
+      if (!(action in sanitizeActions)) {
+        sanitizeActions[action] = []
```
Comment (silvs110) [inline index.html:151]:
```suggestion
        sanitizeActions[action] = [];
```

File: index.html [hunk 2]
```diff
@@ -1,47 +1,221 @@
...
+    }
+
+    const rules = jsyaml.load(localStorage.getItem(fileExtension))["rules"];
+    if (rules == null) {
+      console.log(`Rules not found for "${fileExtension}" file extension.
+                       Check for previous errors or try reloading the page.`)
+      return
+    } else {
+      const viewRulesButton = document.getElementById("view_rules_button");
+      const ruleFilePath = "rules/" + fileExtension + ".yaml"
```
Comment (silvs110) [inline index.html:135]:
```suggestion
      const ruleFilePath = "rules/" + fileExtension + ".yaml";
```

File: index.html [hunk 3]
```diff
@@ -1,47 +1,221 @@
...
+    showSpinner();
+    const fileName = file.name;
+    const fileNameParts = file.name.split('.');
+    const fileExtension = fileNameParts.pop();
+
+    if (!supportedFileExtensions.includes(fileExtension)) {
+      const error = `Unsupported file extension: ${fileExtension}`;
+      console.error(error);
+      alert(error);
+      return
```
Comment (silvs110) [inline index.html:114]:
```suggestion
      return;
```

File: index.html [hunk 4]
```diff
@@ -1,47 +1,221 @@
...
+      // return the next level down
+      return accumulator[key]
+    }, obj)
+    // return the original object
+    return obj
+  };
+
+  for (const supportedFileExtension of supportedFileExtensions) {
+    // Load rules
+    const ruleFilePath = "rules/" + supportedFileExtension + ".yaml"
```
Comment (silvs110) [inline index.html:94]:
```suggestion
    const ruleFilePath = "rules/" + supportedFileExtension + ".yaml";
```

File: index.html [hunk 5]
```diff
@@ -1,47 +1,221 @@
...
+  let sanitizedFileName = null;
+  let supportedFileExtensions = ['har'];
+  let supportedActions = ['remove'];
+
+  const setConfigItem = (path, value, obj) => {
+    // reduce the path array, each iteration dig further into the object properties
+    path.reduce((accumulator, key, i) => {
+      // if you are at the final key set the value
+      if (i === path.length - 1) {
+        accumulator[key] = value
```
Comment (silvs110) [inline index.html:78]:
```suggestion
        accumulator[key] = value;
```

File: index.html [hunk 6]
```diff
@@ -1,47 +1,221 @@
...
+        return accumulator
```
Comment (silvs110) [inline index.html:79]:
```suggestion
        return accumulatorp;
```

**Review by silvs110 — CHANGES_REQUESTED** [review]
Move all the imported js file to a subdirectory of ```script```. Also there was an error when loading a har file  (login into github):
image

File: index.html [hunk 7]
```diff
@@ -1,47 +1,221 @@
...
-      </div>
+      <div id="sanitized_diff_div"></div>
     </div>
+    <div class="card-footer text-body-secondary" id="downloader"></div>
   </div>
+</div>
+
+<!--suppress JSUnresolvedFunction -->
+<script>
+  const diff2HtmlConfiguration = {
```
[the remark this answers is not in the document]
Comment (the-c0d3br34k3r) [reply index.html:56]: Done.

File: index.html [hunk 8]
```diff
@@ -1,47 +1,221 @@
...
+    fetch(ruleFilePath)
+            .then(response => response.text())
+            .then((data) => {
+              localStorage.setItem(supportedFileExtension, data);
+            }).catch(error => console.error(error));
+  }
+
+  async function sanitizeContent(file) {
+    document.getElementById("display_panel").hidden = true;
+    showSpinner();
```
[the remark this answers is not in the document]
Comment (the-c0d3br34k3r) [reply index.html:105]: The spinner does get loaded, it's that the file load and sanitization happen pretty fast so it's not visible.

File: index.html [hunk 9]
```diff
@@ -1,47 +1,221 @@
...
-              <div class="spinner-border text-success" id="spinner"></div>
-            </div>
-          </div>
-        <div class="scrollable" id="scrollable_content">
-          <p class="card-text" id="output_content"></p>
+<div id="display_panel" hidden>
+  <div id="display_card" class="card">
+    <div id="buttons">
+      <input type="button" id="view_rules_button" class="btn btn-secondary" name="view_rules_button"
+             value="View Rules File" onclick="" disabled>
```
[the remark this answers is not in the document]
Comment (the-c0d3br34k3r) [reply index.html:38]: Done

File: index.html (hunk 2 above)
Comment (the-c0d3br34k3r) [reply index.html:135]: Done.

File: index.html (hunk 3 above)
Comment (the-c0d3br34k3r) [reply index.html:114]: Done.

File: index.html (hunk 4 above)
Comment (the-c0d3br34k3r) [reply index.html:94]: Done.

File: index.html (hunk 6 above)
Comment (the-c0d3br34k3r) [reply index.html:79]: Done.

File: index.html [hunk 10]
```diff
@@ -1,47 +1,221 @@
...
+    // reduce the path array, each iteration dig further into the object properties
+    path.reduce((accumulator, key, i) => {
+      // if you are at the final key set the value
+      if (i === path.length - 1) {
+        accumulator[key] = value
+        return accumulator
+      }
+      // test to see if there is a property
+      if (typeof accumulator[key] === 'undefined') {
+        throw new Error('Nothing to see here')
```
[the remark this answers is not in the document]
Comment (the-c0d3br34k3r) [reply index.html:83]: Done.

**Comment by the-c0d3br34k3r** [comment]
> Move all the imported js file to a subdirectory of `script`. Also there was an error when loading a har file (login into github): image

Fixed this issue. Needed to change the JSON path expression to specify `name` as an optional field during lookup.

File: script/script.js [hunk 11]
```diff
@@ -1,41 +1,161 @@
-function getFile(event) {
```
Comment (silvs110) [inline script/script.js:1]: This file is a duplicate of script/lib/script.js

**Review by silvs110 — CHANGES_REQUESTED** [review]
When I click on `View Rules File` in Firefox, it downloads the YAML file instead of opening  a new tab.

File: script/script.js (hunk 11 above)
Comment (the-c0d3br34k3r) [reply script/script.js:1]: Done

**Comment by the-c0d3br34k3r** [comment]
> When I click on `View Rules File` in Firefox, it downloads the YAML file instead of opening a new tab.

Seems like a Firefox thing. As long as the rules file is accessible, it's fine.

**Review by silvs110 — APPROVED** [review]

**Pull request merged**